tests/testthat/test-1-0-0rdlearn_lip.R

```diff
@@ -0,0 +1,26 @@
+test_that("learned cutoffs loop over lip(M)", {
+  # -------------------------------------------------------------------------- #
+  # esimate dif and lip
+  # -------------------------------------------------------------------------- #
+
+  set.seed(12345)
+  result <- rdlearn(y = "elig", x = "saber11", c = "cutoff",
+                    group_name = "department", data = acces2,
+                    fold = 20, M = c(0, 1, 2, 4), cost = 0)
+
+  # -------------------------------------------------------------------------- #
+  # test
+  # -------------------------------------------------------------------------- #
+
+  # Setting tolerance to 5 in case where regrets have ties.
+  data("safecut_M0124C0_test")
```

Comment:
Where do you generate the "ground-truth" data? You might want to check in the script to generate the ground truth data as well: A potential location is `data-raw` (https://r-pkgs.org/data.html#sec-data-data-raw).

Reply:
The goal of test is to check whether the results match those from Yi's original code. In this case, what would be the best approach? Copying and pasting the original code to generate "ground-truth data" within the package might make sense for us but not make sense for users.

Reply:
Got it. Thanks for the clarification.

I think you can add a comment before loading the data that the purpose of this test is to replicate the original result. We can think about more standard unit tests later.

Reply:
For unit tests, I think the basic approach is to generate simulation data and verify whether the package's function can estimate the parameters correctly. 

However, in the case of policy learning, there is no true policy, and I’m struggling to make tests… It seems reasonable to test the intermediate results, but because extrapolation is involved in the process, it’s still difficult. 

Does it sound good to test whether obtained policy improves regret compared to a baseline instead? Or are there any other good way to test ?

And, if you know of any policy learning packages that could serve as a useful reference, please let me know. 
I looked into how the package for [Athey and Wager’s work](https://onlinelibrary.wiley.com/doi/abs/10.3982/ECTA15732) and its [package](https://cran.r-project.org/src/contrib/Archive/policytree/), but I couldn’t find any helpful hints…

Reply:
@carolyyzz In the section on simulation studies of the paper, it mentions “... and then further calibrating the fitted coefficients such that the baseline policy π̃ equals the oracle policy π∗ under Scenario A.” Could you explain how this calibration was performed in detail? It might help in constructing tests. Also, if you have any additional ideas for creating tests, I’d greatly appreciate it.

Reply:
> For unit tests, I think the basic approach is to generate simulation data and verify whether the package's function can estimate the parameters correctly.
> 
> However, in the case of policy learning, there is no true policy, and I’m struggling to make tests… It seems reasonable to test the intermediate results, but because extrapolation is involved in the process, it’s still difficult.
> 

I think replicating the original results kind of serves the purpose of testing the correctness of overall implementation (of course assuming the original data is generated correctly).

For unit tests (not urgent so we can keep discussing), we can think about testing individual functions separately. For example, we can test if the [`check_input`](https://github.com/kkawato/rdlearn/blob/main/R/1-1-0check_input.R#L19C1-L19C12) function correctly throws an error when we provide unsupported inputs, or the [`calculate_regret`](https://github.com/kkawato/rdlearn/blob/main/R/1-4-2calculate_regret.R#L16) returns the values that correspond to an analytical result (given the math we should be able to analytically compute the regret in some specific cases), etc.

Reply:
> @carolyyzz In the section on simulation studies of the paper, it mentions “... and then further calibrating the fitted coefficients such that the baseline policy π̃ equals the oracle policy π∗ under Scenario A.” Could you explain how this calibration was performed in detail? It might help in constructing tests. Also, if you have any additional ideas for creating tests, I’d greatly appreciate it.

We first fit a third-order polynomial of $g(X)$ and obtained the fitted coefficients $\gamma_0$ and $\gamma_1$. Next, we observed that the two outcome functions for the treated and control groups,

$$
\gamma_1^\top g(X) \quad \text{and} \quad \gamma_0^\top g(X),
$$

intersect at a point roughly in the middle between the two baseline cutoffs,

$$
-571 \quad \text{and} \quad -850.
$$

To align the baseline policy $\tilde{\pi}$ with the oracle policy $\pi^*$, we shifted both outcome lines to the right so that they now intersect at

$$
\tilde{c}_2 = -571.
$$

This adjustment ensures that:

$$
\gamma_1^\top g(X) \geq \gamma_0^\top g(X) \quad \text{for} \quad X \geq -571,
$$

and

$$
\gamma_1^\top g(X) < \gamma_0^\top g(X) \quad \text{for} \quad X < -571.
$$

As a result, the baseline policy aligns with the oracle optimal policy for group 2.

For group 1, we also tuned the parameter in the $d(w, X)$ function to ensure that:

$$
\gamma_1^\top g(X) - d(1, X) > \gamma_0^\top g(X) - d(0, X) \quad \text{for} \quad X > -850.
$$

Since the candidate cutoff must satisfy:

$$
X > -850,
$$

there is no gain in increasing the cutoff. This ensures that the baseline cutoff for group 1 also corresponds to its oracle optimal policy.

Reply:
Thank you! I made a unit test `test-simulation.R` based on this simulation study to check whether `rdlearn` correctly outputs the oracle policy (baseline cutoff).